fw-autotune: add amplitude detection states

mahima-yoga opened the pull request:
### Solved Problem

In fixed-wing autotune, the system identification signal amplitude is currently set via the user-defined parameter `FW_AT_SYSID_AMP`. This can lead to excessively high angular rates. 

Specifically for yaw and pitch, these high angular rates can lead to non-linear behaviour. This can result is bad system identification. 

### Solution
Increases signal ampltiude on a 1Hz sine input until a target angular rate is reached per axis. Identification signal is then scaled with this amplitude instead of the user-defined parameter. After testing, rates have been set to: 

1. Roll rate: 0.8 rad/s (approx 45 deg/s) 
2. Pitch rate: 0.5 rad/s (approx 30 deg/s) 
3. Yaw rate: 0.5 rad/s (approx 30 deg/s) 

**Steps:**

1. Apply a 1 Hz sine input to the axis being tuned.
4. Measure the angular rate over one period:
  (i) If the target angular rate is reached, verify consistency by observing a second period.
5. If the target rate is not reached after the first period, or not verified in the second:
  (i) Increase the signal amplitude
  (ii) Repeat the process from step 2
6. If the target angular rate is verified in step 2:
  (i) Take the amplitude from the previous run to ensure we don't exceed the target 

This amplitude is then applied to the system identification input signal. 

Additional Notes:
- System identification runs during this state

### Changelog Entry
For release notes:
```
Feature/Bugfix detect and limit amplitude of fixed-wing autotune ID maneuver to 0.8/0.5/0.5 rad/s for roll/pitch/yaw. 
```
### Test coverage

**SITL TEST:** 

<img width="1900" height="957" alt="image" src="https://github.com/user-attachments/assets/80019ece-7fa9-4e0d-b5b3-2adea9560dc8" />


**FLIGHT TESTS:**
 

Tested on 3 different platforms. In each flight, different target rates were used to identify what is sufficient for system ID. The figures below show ID results for the chosen target rates. Online identification is also still consistent. 

1. [T1-ranger](https://review.px4.io/plot_app?log=fed7bc7b-906d-442e-9ad0-1b881344a30e): no rudder. 

<img width="1302" height="764" alt="image" src="https://github.com/user-attachments/assets/cc47b39b-f6fe-4502-8c32-c1e871dc1638" />

<img width="1302" height="764" alt="image" src="https://github.com/user-attachments/assets/0d552383-e7c9-454a-b3d3-30ab24102f6c" />

2. [Zohd Altus ](https://review.px4.io/plot_app?log=695c15be-c7f0-4c64-b6dc-05ea08463195)

<img width="1302" height="764" alt="image" src="https://github.com/user-attachments/assets/9b879f7f-d035-4745-affe-805d53b0b1e8" />
<img width="1302" height="764" alt="image" src="https://github.com/user-attachments/assets/8f3951ba-1750-4496-9be5-dabb2d96cd91" />
<img width="1302" height="764" alt="image" src="https://github.com/user-attachments/assets/444cfce9-dd82-468e-baa1-cc7c46a6a6bd" />


4. [Loong](https://review.px4.io/plot_app?log=74e66ca7-b9fc-4b84-929b-765920f86774) (VTOL) 
<img width="1302" height="764" alt="image" src="https://github.com/user-attachments/assets/2324dd85-b3cb-4e56-b688-6fb07b175a1b" />

<img width="1302" height="764" alt="image" src="https://github.com/user-attachments/assets/6aeb20d3-ce36-4750-a4f4-b59a12f42881" />

bresch commented on src/modules/fw_autotune_attitude_control/fw_autotune_attitude_control.hpp:
```diff
@@ -149,6 +156,19 @@ class FwAutotuneAttitudeControl : public ModuleBase<FwAutotuneAttitudeControl>,
+	float _signal_amp_step{0.1f};
+	float _target_rate{0.5f};
+	hrt_abstime _time_last_amplitude_increase{0};
+	float _max_measured_rate_T1{0.0f};
+	float _max_measured_rate_T2{0.0f};
+	bool _rate_reached_T1{false};
+	bool _rate_reached_T2{false};
```
You don't really need to store all of this.
For example, the max measured rate is not useful as you just need to know if the current rate is above the threshold. Once reached, you can set `_rate_reached_T1`, wait for T2 to start and check a second time. If reached again, directly execute what needs to be done.

mahima-yoga replied:
@bresch Fair I can remove the `_max_measured_rate_Tx`.  The only argument I would have to keep `_rate_reached_T2` (and not directly execute the next steps) would be to avoid having to compute the phase shift between the amplitude detection signal and the identification signal.  If you always let the full period play out the signals can align nicely without much else overhead. 

(Unless you see a simpler way to do this / value in having a function that can align phase)

bresch commented on src/modules/fw_autotune_attitude_control/fw_autotune_attitude_control.hpp:
```diff
@@ -149,6 +156,19 @@ class FwAutotuneAttitudeControl : public ModuleBase<FwAutotuneAttitudeControl>,
+	bool _rate_reached_T1{false};
+	bool _rate_reached_T2{false};
+
+	vehicle_angular_velocity_s _angular_velocity{};
```
if you just need the xyz part of the topic, just keep that part in a `Vector3f`

bresch commented on src/modules/fw_autotune_attitude_control/fw_autotune_attitude_control.cpp:
```diff
@@ -354,12 +369,42 @@ void FwAutotuneAttitudeControl::updateStateMachine(hrt_abstime now)
+				break;
+			}
+
+			const hrt_abstime dt = (now - _time_last_amplitude_increase);
```
calling this `dt` is not wrong, but it can be confusing as this is usually the time between 2 frames. Here, you measure the time since last amplitude increase; so `time_since_last_amplitude_increase` ?

mahima-yoga replied:
good point, addressed

bresch reviewed (changes requested):
The algorithm looks good, but the implementation could be simplified quite a bit

mahima-yoga commented on src/modules/fw_autotune_attitude_control/fw_autotune_attitude_control.cpp:
```diff
@@ -308,26 +323,40 @@ void FwAutotuneAttitudeControl::updateStateMachine(hrt_abstime now)
-			_signal_sign = 1;
-			_input_scale = 1.f / _param_fw_rr_p.get();
 			_signal_filter.reset(0.f);
-			_gains_backup_available = false;
```
Was accidentally removed - added it back in.

sfuhrer commented on docs/en/config/_autotune.md:
```diff
@@ -170,9 +170,19 @@ Fast oscillations (more than 1 oscillation per second): this is because the gain
+By default, the autotune maneuvers ensure that a sufficient angular rate is reached for system identification. If the signal-to-noise ratio of the vehicle is low, the system identification algorithm might have issues finding the correct coefficients. Ensure that there is no excessive noise and/or platform vibration.
+
+</div>
+
```
Worth specifying the expected rates here? A screenshot like you put into the PR description would be nice as well, with the amplitude increase noticeable.

sfuhrer reviewed (dismissed):
Thanks, looks good!

mahima-yoga committed:
fw-autotune: detect and limit amplitude of ID maneuver
Increases signal ampltiude on a 1Hz sin input until a target rate (R/P/Y = 0.8/0.5/0.5 rad/s) is reached. Identification signal is then scaled with this ampltitude instead of the user-defined parameter.

hamishwillee commented on docs/en/config/_autotune.md:
```diff
@@ -104,6 +112,10 @@ The test steps are:
 5. The tuning will be immediately/automatically be applied and tested in flight (by default).
    PX4 will then run a 4 second test and revert the new tuning if a problem is detected.
 
+The figure below shows how steps 4 and 5 might look in flight on the pitch axis. The pitch rate gradually increases up until it reaches the target. This amplitude is then held while the signal frequency is increased. You can then see how the tuned system is able to follow the setpoint in the test signal.
```
```suggestion
The figure below shows how steps 4 and 5 might look in flight on the pitch axis.
The pitch rate gradually increases up until it reaches the target.
This amplitude is then held while the signal frequency is increased.
You can then see how the tuned system is able to follow the setpoint in the test signal.
```

hamishwillee replied:
PS Other docs look good!

hamishwillee commented on msg/AutotuneAttitudeControlStatus.msg:
```diff
@@ -1,8 +1,10 @@
-uint64 timestamp                # time since system start (microseconds)
+# Autotune attitude control status
```
@mahima-yoga While we're here can we add documentation in the standard form for this message please? After all, you're now the expert and no one will be able to do it better.

I have started it - for inspiration see https://docs.px4.io/main/en/msg_docs/ArmingCheckReply.html

I'd hope for a more complete description below this one explaining what publishes and subscribes it, and anything that can't be inferred from the name that would help developers maintain it and know when they should touch it.

hamishwillee commented on msg/AutotuneAttitudeControlStatus.msg:
```diff
@@ -1,8 +1,10 @@
-float32 fitness                 # fitness of the parameter estimate
+uint64 timestamp  # [us] Time since system start
+
+float32[5] coeff      # [-] Coefficients of the identified discrete-time model
```
FYI, 
- `[x]` means "units", and `[-]` specifically means "no units". Would hope for units for the other fields as well.
- Comment indentation is done by block of field and aligns two spaces after the longest field name.

hamishwillee replied:
@mahima-yoga Thanks for the update. Looks great. Resolving!

hamishwillee commented on msg/AutotuneAttitudeControlStatus.msg:
```diff
@@ -17,19 +19,21 @@ float32[3] rate_sp
 float32 y_filt
 
-uint8 STATE_IDLE = 0
+uint8 state                               # [@enum STATE]
```
FYI `[@enum Xxx]` identifies a prefix for the values that can be used with the field. By convention the values are listed after the field. 
If you can provide useful documentation for these then great. Especially if there is anything that isn't obvious from the name.

hamishwillee reviewed (approved):
Approving for the docs. Thank you!

mahima-yoga committed:
docs: update fixed-wing autotune docs

mahima-yoga committed:
docs: add description to AutotuneAttitudeControlStatus.msg
